### DISCUSSION TITLE
Proposal:  Anonymous Lambda Creation

### DISCUSSION THREAD

**Discussion by iam3yal** [dotnet/csharplang#148] [General]
## Summary
[summary]: #summary

Allow lambdas to be created "without" assigning them to variables.

## Motivation
[motivation]: #motivation

Many times, especially in my unit tests I have extension methods that operate on variations of the types `Action<>` and/or `Func<>`, atm I need to create lambdas and assign them to a temp variable and only then I can call these extension methods as opposed to calling them directly on the lambda expression, it looks like this:

```
Action action = () => XmlCommandRootElementConverter.ToCommandModelCollection(null);
action.ShouldThrow<ArgumentNullException>()
```

The temp variable is completely redundant here so with this proposal we could do the following:

```
(() => XmlCommandRootElementConverter.ToCommandModelCollection(null))
    .ShouldThrow<ArgumentNullException>();
````

## Detailed design
[design]: #detailed-design

Today, the parenthesis operator is mostly used for casting and function calls so what I'm proposing is to extend this to allow lambdas to be created "without" assigning them to variables.

The compiler would probably need to create this temp variable behind the scene but from the language standpoint, the programmer wouldn't have to declare it.

**Example 1:**

```
(() => Console.WriteLine("Hello World"));
```

Equivalent to:

```
Action action = () => Console.WriteLine("Hello World"); 
```

**Example 2:**

```
(() => "Hello World");
```

Equivalent to:

```
Func<string> func = () => "Hello World";
```

**Example 3:** (thanks to @jnm2) 

```
// On the UI thread.
await Task.WhenAll(
    (async () =>
    {
        var foo = await persistenceProvider.GetFoo();
        UpdateFoo(foo.X, foo.Y); // Update UI with foo without waiting for bar
    })(),
    (async () =>
    {
        var bar = await persistenceProvider.GetBar();
        UpdateFoo(bar.X, bar.Y, bar.Z); // Update UI with bar without waiting for foo
    })());

// foo and bar are both finished
```

## Drawbacks
[drawbacks]: #drawbacks

Don't think there's any drawbacks.

## Alternatives
[alternatives]: #alternatives

1. Declare a variable, create a lambda and assign it to a temp variable.

2. Create a method with a specific parameter that matches the delegate type you want to use and then pass the lambda to it.

    The drawback of creating "dummy" methods such as `Do(() => Console.WriteLine("Hello World"));` is that it can be painful to maintain if you use many variations of these delegate types and sometimes you can't even use 3rd-party libs, depends where you work and/or licenses.

    Another drawback is that both of the approaches above add redundancy and verbosity and as such they may decrease the legibility of the code.

3. Create an instance of the delegate type directly.

    ```
    new Action<string>(Console.WriteLine).MyExtMethod();
    ```

### Known Issues:

* There's some ambiguity in the examples and today the compiler cannot tell the difference so in order to disambiguate, it would need to know about the various delegate types it needs to infer to based on the context.

### Related Topics:

* [#150](https://github.com/dotnet/csharplang/issues/150)
* [#149](https://github.com/dotnet/csharplang/issues/149)
* [#16105](https://github.com/dotnet/roslyn/issues/16105)

**Comment by svick** [comment]
I don't understand the motivation here. *Why* would you want to create lambda just to invoke it immediately? Your examples either don't make sense to me (statements that create a function, but never invoke it) or could be simplified by removing the lambda.

Also, why do you need new syntax for this in the first place? If you have a helper function like this:

```c#
static T Invoke<T>(Func<T> f) => f();
```

Then you can write Example 4 like this:

```c#
Console.WriteLine(Invoke(() => "foo"));
```

(Your `Foo` method might be the same idea.)

If it's just about saving some typing, then to justify new syntax, I think this would have to be really common, which I don't see how it would be.

**Comment by iam3yal** [comment]
@svick Many times, especially in my unit tests I have extension methods that operates on variations of `Action<>` and `Func<>`, atm I need to invoke the lambda and assign the result to a temp variable and then call these extension methods, it looks like this:

```
Action action = () => XmlCommandRootElementConverter.ToCommandModelCollection(null);
action.ShouldThrow<ArgumentNullException>()
```

The temp variable is completely redundant here so with this proposal you would do this:

```
(() => XmlCommandRootElementConverter.ToCommandModelCollection(null))
    .ShouldThrow<ArgumentNullException>();
```

Today, 3rd party libs offer helper methods to workaround it but depending on a library just to allow this is a bit too much, sometimes not even possible and creating these helper methods in every project can be painful in terms of maintenance and sometimes it might not be immediately clear what is the purpose of the method so it can actually decrease the legibility of the tests.

**Comment by jnm2** [comment]
Ah, that makes so much more sense. You want delegate extension methods to be available on method groups/lambdas with compatible signatures.

**Comment by svick** [comment]
@eyalsk I didn't realize this issue is copied from the Roslyn repo: https://github.com/dotnet/roslyn/issues/14401, where some of this discussion already happened.

And there's also a proposal specifically about calling extension methods on lambdas: https://github.com/dotnet/roslyn/issues/16105.

**Comment by iam3yal** [comment]
@svick Yeah, I just copied some of the things I thought that were important but I'll revisit this proposal a bit to make it crystal clear.

**Comment by iam3yal** [comment]
@jnm2, @svick I revised the post, I also wrote invoke in few places where I really meant create.

**Comment by jnm2** [comment]
I still see no benefit to examples 1 thru 4. The example you didn't list in the post, using extension method on lambda without explicitly creating a delegate, is the only thing that makes sense here to me.

**Comment by iam3yal** [comment]
@jnm2 Hmm so you're saying you want me to change the examples to something like this?

Example 1:
```
(() => Console.WriteLine("Hello World")).SomeExtMethod();
```
Equivalent to:
```
Action action = () => Console.WriteLine("Hello World"); 
action.SomeExtMethod();
```

**Comment by jnm2** [comment]
@eyalsk Yes, that's the only one where I see value. Unless I'm missing something.

There's no reason to do this as a statement: `(() => Console.WriteLine("Hello World"));` Creating a delegate has no side effects.

You'd be passing it to a parameter which is already typed as a specific delegate, or you'd do `var foo = (() => Console.WriteLine("Hello World"));` which is arguably not an improvement over `Action foo = (() => Console.WriteLine("Hello World"));`.

There's no reason to do `(() => "Hello World")()` over `"Hello World"`.
There's no reason to do `(() => { var scoped = "Hello World"; Console.WriteLine(scoped); })()` over `{ var scoped = "Hello World"; Console.WriteLine(scoped); }`.

**Comment by iam3yal** [comment]
@jnm2 Okay, these are great points.

I listed all of these examples for completeness but sure I'll modify the post and remove examples 2 and 4.

Thanks. :)

**Comment by vladd** [comment]
One possible way for "self-invoking" lambda is this: `((Action)(() => { }))();`

**Comment by iam3yal** [comment]
@vladd Yes but then why not `new Action(() => { })();`? :)

**Comment by jnm2** [comment]
And then, why not `{ }`? ;)

**Comment by gafter** [comment]
I believe the proposal here is to modify fourth bullet, below, in the specification for [Extension method invocations](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#extension-method-invocations) from

> An extension method `Ci.Mj` is ***eligible*** if:
> 
> *  `Ci` is a non-generic, non-nested class
> *  The name of `Mj` is *identifier*
> *  `Mj` is accessible and applicable when applied to the arguments as a static method as shown above
> *  An implicit identity, reference or boxing conversion exists from *expr* to the type of the first parameter of `Mj`.

to

> An extension method `Ci.Mj` is ***eligible*** if:
> 
> *  `Ci` is a non-generic, non-nested class
> *  The name of `Mj` is *identifier*
> *  `Mj` is accessible and applicable when applied to the arguments as a static method as shown above
> *  An implicit identity, reference, boxing **or lambda** conversion exists from *expr* to the type of the first parameter of `Mj`.

I've also heard suggestions recently to include method group conversions and tuple conversions as well.

**Comment by iam3yal** [comment]
@gafter Yes, this is exactly what I was trying to convey.

**Comment by jnm2** [comment]
@eyalsk Oooh, I just ran into a case where I actually do want the syntax you first proposed, but only because it's async:

```c#
// On the UI thread.
await Task.WhenAll(
    (async () =>
    {
        var foo = await persistenceProvider.GetFoo();
        UpdateFoo(foo.X, foo.Y); // Update UI with foo without waiting for bar
    })(),
    (async () =>
    {
        var bar = await persistenceProvider.GetBar();
        UpdateFoo(bar.X, bar.Y, bar.Z); // Update UI with bar without waiting for foo
    })());

// foo and bar are both finished
```

You can't simply replace this with the `{ }` contents. I'll have to take back what I said!

Now inserting `new Func<Task>` twice will not hurt the readability, which is already low, so I'm not sure if this is really an example worth adding. But, it shows I was wrong about always being able to replace the IIFE with its body. That's not true if a method builder is applied to the lambda. If we had anonymous iterators the same thing would apply.

Plus, this is personal, but I really dislike invoking delegates without writing `.Invoke`.

**Comment by svick** [comment]
@jnm2 I think that is a good use case for this, though using local functions for the same purpose is not that bad either:

```c#
async Task GetAndUpdateFoo()
{
    var foo = await persistenceProvider.GetFoo();
    UpdateFoo(foo.X, foo.Y); // Update UI with foo without waiting for bar
}

async Task GetAndUpdateBar()
{
    var bar = await persistenceProvider.GetBar();
    UpdateFoo(bar.X, bar.Y, bar.Z); // Update UI with bar without waiting for foo
}

// On the UI thread.
await Task.WhenAll(GetAndUpdateFoo(), GetAndUpdateBar());
// foo and bar are both finished
```

This way is more verbose, but it's also less of a syntax soup. (And I think it also makes the likely bug of calling `UpdateFoo` in the `bar` lambda more visible, in case the compiler didn't catch it. :wink:)

**Comment by jnm2** [comment]
> And I think it also makes the likely bug of calling UpdateFoo in the bar lambda more visible, in case the compiler didn't catch it.

Hehe, that was meant to be an imaginary `UpdateBar` method. Yay for ad-hoc ```` ```c#```` blocks :'D

**Comment by iam3yal** [comment]
@jnm2 Haha.. I'll add it. ;)

**Comment by Thaina** [comment]
Well I would say I want anonymous local function. And so I support this feature

**Comment by NetMage** [comment]
Wouldn't the change in extension methods allow you to create your own set of `Invoke` extension methods?

    public static T Invoke<T>(this Func<T> f) => f();

    var ans = (() => 7).Invoke();

**Comment by iam3yal** [comment]
@NetMage No because what you're proposing has exactly the same drawbacks as the alternative section paragraph 2 that I've written above. <s>only now, you declare a delegate instead of a method.</s>

**Comment by NetMage** [comment]
Given that Invoke can be a generic method I don't think it would be any worse than Func<> has turned out to be.

**Comment by iam3yal** [comment]
@NetMage I don't think that just because the framework provides variations to these types which is basically set of standard types that everyone use today, people need to follow and do that for their own  codebases, just because it's the correct solution for a framework doesn't mean it's the correct one for any kind of software and if the language can save us from writing redundant code then there's no reason not to do it, at least in this case.

**Comment by ghost** [comment]
@eyalsk 

> Allow lambdas to be created "without" assigning them to variables.

> Create a method with a specific parameter that matches the delegate type you want to use and then pass the lambda to it. [...]

What do you think about an generalized lambda factory approach?

```c#
New.Action(() => Console.WriteLine("Hello World")).Invoke();
/* vs. */
(() => Console.WriteLine("Hello World")).Invoke();

//Do(() => Console.WriteLine("Hello World"));
```
Which is suitable for type inference too

```c#
var a = New.Func((int x) => x);
var b = New.Func((int x) => x == 42);
			
//var a = (Func<int, int>)(x => x);
//var b = (Func<int, bool>)(x => x == 42);

var handler = New.EventHandler((sender, args) => WriteLine("Hello"));
```
Implementation
```c#
public class New
{
    public static EventHandler EventHandler(EventHandler h) => h;
    public static EventHandler<T> EventHandler<T>(EventHandler<T> h) where T : EventArgs => h;

    public static Func<Z> Func<Z>(Func<Z> f) => f;
    public static Func<A, Z> Func<A, Z>(Func<A, Z> f) => f;
    public static Func<A, B, Z> Func<A, B, Z>(Func<A, B, Z> f) => f;
    // ...
    
    public static Action Action(Action a) => a;
    public static Action<A> Action<A>(Action<A> a) => a;
    public static Action<A, B> Action<A, B>(Action<A, B> a) => a;
    // ...
}
```
Posted by @Makeman-from-Makeloft
